Port WKWebExtensionLocalization to C++

ItsJamie9494 opened the pull request:
#### 944567bb72da80b8a98ac4aac76f70b700373989
<pre>
Port WKWebExtensionLocalization to C++
<a href="https://webkit.org/b/283791">https://webkit.org/b/283791</a>

Reviewed by Timothy Hatcher.

Create a new WebExtensionLocalization shared class and replace usage of WKWebExtensionLocalization with the new class wherever possible.

* Source/WTF/wtf/JSONValues.h:
* Source/WebKit/Shared/Extensions/WebExtensionContextParameters.h:
* Source/WebKit/Shared/Extensions/WebExtensionContextParameters.serialization.in:
* Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp: Added.
(WebKit::WebExtensionLocalization::WebExtensionLocalization):
(WebKit::WebExtensionLocalization::loadRegionalLocalization):
(WebKit::WebExtensionLocalization::localizedJSONforJSON):
(WebKit::WebExtensionLocalization::localizedStringForKey):
(WebKit::WebExtensionLocalization::localizedArrayForArray):
(WebKit::WebExtensionLocalization::localizedStringForString):
(WebKit::WebExtensionLocalization::localizationJSONForWebExtension):
(WebKit::WebExtensionLocalization::predefinedMessages):
(WebKit::WebExtensionLocalization::stringByReplacingNamedPlaceholdersInString):
(WebKit::WebExtensionLocalization::stringByReplacingPositionalPlaceholdersInString):
* Source/WebKit/Shared/Extensions/WebExtensionLocalization.h: Added.
(WebKit::WebExtensionLocalization::create):
(WebKit::WebExtensionLocalization::uniqueIdentifier):
(WebKit::WebExtensionLocalization::localizationJSON):
(WebKit::WebExtensionLocalization::localizedStringForKey):
(WebKit::WebExtensionLocalization::loadRegionalLocalization):
(WebKit::WebExtensionLocalization::stringByReplacingPositionalPlaceholdersInString):
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.cpp:
(WebKit::JSONWithLowercaseKeys):
(WebKit::mergeJSON):
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.h:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::parseManifest):
(WebKit::WebExtension::serializeLocalization): Deleted.
(WebKit::WebExtension::localization): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::localization):
(WebKit::WebExtensionContext::localizedResourceString):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionURLSchemeHandlerCocoa.mm:
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
(WebKit::WebExtension::serializeLocalization):
(WebKit::WebExtension::localization):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPILocalizationCocoa.mm:
(WebKit::WebExtensionAPILocalization::getMessage):
* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionContextProxyCocoa.mm:
(WebKit::WebExtensionContextProxy::getOrCreate):
(WebKit::WebExtensionContextProxy::parseLocalization):
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:

Canonical link: <a href="https://commits.webkit.org/287187@main">https://commits.webkit.org/287187@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/04d703a7cf9132d124a03ff28433f935b94c33c4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78734 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57778 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32116 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83394 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29996 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/80867 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66930 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6059 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61663 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19586 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81801 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51695 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/70988 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41971 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49041 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/25721 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28336 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/71875 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70149 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26104 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84763 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/77967 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6099 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4216 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69888 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6261 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67679 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69142 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13181 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/11785 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/100279 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12149 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6044 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/11953 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21906 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6030 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9466 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7818 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WTF/wtf/JSONValues.h:
```diff
@@ -211,7 +211,7 @@ class SUPPRESS_REFCOUNTED_WITHOUT_VIRTUAL_DESTRUCTOR ObjectBase : public Value {
     const_iterator end() const { return m_map.end(); }
 
-    DataStorage::KeysConstIteratorRange keys() const { return m_map.keys(); }
+    OrderStorage keys() const { return m_order; }
```
Nice!

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionContextParameters.h:
```diff
@@ -45,7 +45,7 @@ struct WebExtensionContextParameters {
     HashMap<String, WallTime> grantedPermissions;
 
-    Ref<API::Data> localizationJSON;
+    RefPtr<API::Data> localizationJSON;
```
Why can't it stay Ref to signal it is never missing / null?

ItsJamie9494 replied:
It can be null, in `serializeLocalization` if the localization class isn't created or there's no localization JSON saved, the function returns nullptr, and I beleive using something like `{ }` didn't work initially.

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionContextParameters.serialization.in:
```diff
@@ -31,7 +31,7 @@ struct WebKit::WebExtensionContextParameters {
     HashMap<String, WallTime> grantedPermissions;
 
-    Ref<API::Data> localizationJSON;
+    RefPtr<API::Data> localizationJSON;
```
Ditto.

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp:
```diff
@@ -0,0 +1,349 @@
+        return;
+    }
+
+    auto defaultLocaleJSON = localizationJSONForWebExtension(webExtension, defaultLocaleString);
```
```suggestion
    RefPtr defaultLocaleJSON = localizationJSONForWebExtension(webExtension, defaultLocaleString);
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp:
```diff
@@ -0,0 +1,349 @@
+    ASSERT(!uniqueIdentifier.isEmpty());
+    ASSERT(localizedJSON);
+
+    String localeString = emptyString();
```
```suggestion
    auto localeString = emptyString();
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp:
```diff
@@ -0,0 +1,349 @@
+    ASSERT(localizedJSON);
+
+    String localeString = emptyString();
+    if (auto predefinedMessages = localizedJSON->getObject(predefinedMessageUILocale))
```
```suggestion
    if (RefPtr predefinedMessages = localizedJSON->getObject(predefinedMessageUILocale))
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp:
```diff
@@ -0,0 +1,349 @@
+    if (!json || !json->size())
+        return;
+
+    for (auto key : json->keys()) {
```
```suggestion
    for (auto& key : json->keys()) {
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp:
```diff
@@ -0,0 +1,349 @@
+    if (!localizationJSON)
+        return emptyString();
+
+    RefPtr<JSON::Object> stringJSON = localizationJSON->getObject(key.convertToASCIILowercase());
```
```suggestion
    RefPtr stringJSON = localizationJSON->getObject(key.convertToASCIILowercase());
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp:
```diff
@@ -0,0 +1,349 @@
+    if (!stringJSON)
+        return emptyString();
+
+    String localizedString = stringJSON->getString(messageKey);
```
```suggestion
    auto localizedString = stringJSON->getString(messageKey);
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp:
```diff
@@ -0,0 +1,349 @@
+    if (!localizedString.length())
+        return emptyString();
+
+    RefPtr<JSON::Object> namedPlaceholders = JSONWithLowercaseKeys(stringJSON->getObject(placeholdersKey));
```
```suggestion
    RefPtr namedPlaceholders = JSONWithLowercaseKeys(stringJSON->getObject(placeholdersKey));
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp:
```diff
@@ -0,0 +1,349 @@
+    if (!json)
+        return;
+
+    for (auto value : *json) {
```
```suggestion
    for (RefPtr value : *json) {
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp:
```diff
@@ -0,0 +1,349 @@
+{
+    String localizedString = sourceString;
+
+    JSC::Yarr::RegularExpression localizableStringRegularExpression = JSC::Yarr::RegularExpression("__MSG_([A-Za-z0-9_@]+)__"_s, { });
```
```suggestion
    auto localizableStringRegularExpression = JSC::Yarr::RegularExpression("__MSG_([A-Za-z0-9_@]+)__"_s, { });
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp:
```diff
@@ -0,0 +1,349 @@
+            break;
+
+        String key = localizedString.substring(index, matchLength);
+        String localizedReplacement = localizedStringForKey(key);
```
```suggestion
        auto key = localizedString.substring(index, matchLength);
        auto localizedReplacement = localizedStringForKey(key);
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp:
```diff
@@ -0,0 +1,349 @@
+    pathBuilder.append("_locales/"_s);
+    pathBuilder.append(withLocale);
+    pathBuilder.append("/messages.json"_s);
+    String path = pathBuilder.toString();
```
```suggestion
    pathBuilder.append("/messages.json"_s);

    auto path = pathBuilder.toString();
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp:
```diff
@@ -0,0 +1,349 @@
+        return nullptr;
+    }
+
+    return JSON::Value::parseJSON(String::fromUTF8(data->span()))->asObject();
```
`parseJSON()` can return null. That should be checked before calling `asObject()`.

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp:
```diff
@@ -0,0 +1,349 @@
+
+Ref<JSON::Object> WebExtensionLocalization::predefinedMessages(RefPtr<JSON::Object> localizedJSON)
+{
+    Ref<JSON::Object> predefinedMessages = JSON::Object::create();
```
```suggestion
    Ref predefinedMessages = JSON::Object::create();
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp:
```diff
@@ -0,0 +1,349 @@
+    };
+
+    if (localizedJSON && !m_localeString.isEmpty()) {
+        if (auto firstMessageObject = localizedJSON->getObject(*(localizedJSON->keys().begin()))) {
```
```suggestion
        if (RefPtr firstMessageObject = localizedJSON->getObject(*(localizedJSON->keys().begin()))) {
```

You should check `localizedJSON->keys()` is not empty, as this would fail if it was.

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp:
```diff
@@ -0,0 +1,349 @@
+        if (auto firstMessageObject = localizedJSON->getObject(*(localizedJSON->keys().begin()))) {
+            auto firstMessage = firstMessageObject->getString(messageKey);
+
+            if (auto writingDirection = firstMessage.defaultWritingDirection(); writingDirection == U_LEFT_TO_RIGHT) {
```
While this is clever, this will not always be accurate. If the messages have a mix of content, and not always LTR, etc, then this will be wrong.

We should extend `WebCore::Locale` and have `LocaleCocoa` use `[NSParagraphStyle defaultWritingDirectionForLanguage:]` for this and whatever is needed on other platforms for the correct answer based solely on the locale string.

ItsJamie9494 replied:
As far as I can tell, WebCore::Locale isn't public to libraries outside WebCore, so it needs to be made public first (I believe this should just be adding the headerfiles to Headers.cmake + marking them as Private in xcode but that doesn't seem to expose them)

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp:
```diff
@@ -0,0 +1,349 @@
+{
+    String localizedString = sourceString;
+
+    JSC::Yarr::RegularExpression localizableStringRegularExpression = JSC::Yarr::RegularExpression("(?:[^$]|^)(\\$([A-Za-z0-9_@]+)\\$)"_s, { });
```
```suggestion
    auto localizableStringRegularExpression = JSC::Yarr::RegularExpression("(?:[^$]|^)(\\$([A-Za-z0-9_@]+)\\$)"_s, { });
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp:
```diff
@@ -0,0 +1,349 @@
+        if (index < 0)
+            break;
+
+        String originalKey = localizedString.substring(index, matchLength);
```
```suggestion
        auto originalKey = localizedString.substring(index, matchLength);
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp:
```diff
@@ -0,0 +1,349 @@
+
+        String key = originalKey.trim(isASCIIWhitespace).substring(1, originalKey.length() - 2).convertToASCIILowercase();
+
+        String localizedReplacement = placeholders->getObject(key) ? placeholders->getObject(key)->getString(placeholderDictionaryContentKey) : emptyString();
```
```suggestion
        auto key = originalKey.trim(isASCIIWhitespace).substring(1, originalKey.length() - 2).convertToASCIILowercase();

        auto localizedReplacement = placeholders->getObject(key) ? placeholders->getObject(key)->getString(placeholderDictionaryContentKey) : emptyString();
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp:
```diff
@@ -0,0 +1,349 @@
+{
+    String localizedString = sourceString;
+
+    JSC::Yarr::RegularExpression localizableStringRegularExpression = JSC::Yarr::RegularExpression("(?:[^$]|^)(\\$([0-9]))"_s, { });
```
```suggestion
    auto localizedString = sourceString;

    auto localizableStringRegularExpression = JSC::Yarr::RegularExpression("(?:[^$]|^)(\\$([0-9]))"_s, { });
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp:
```diff
@@ -0,0 +1,349 @@
+            break;
+
+        String originalKey = localizedString.substring(index, matchLength).trim(isASCIIWhitespace<UChar>);
+        String key = originalKey.substring(1, originalKey.length());
```
```suggestion
        auto originalKey = localizedString.substring(index, matchLength).trim(isASCIIWhitespace<UChar>);
        auto key = originalKey.substring(1, originalKey.length());
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionUtilities.cpp:
```diff
@@ -72,6 +72,32 @@ double largestDisplayScale()
+    if (!jsonA || !jsonB)
+        return jsonA ?: jsonB;
+
+    RefPtr<JSON::Object> mergedObject = jsonA.copyRef();
```
```suggestion
    RefPtr mergedObject = jsonA.copyRef();
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionUtilities.cpp:
```diff
@@ -72,6 +72,32 @@ double largestDisplayScale()
+    if (!json)
+        return json;
+
+    Ref<JSON::Object> newObject = JSON::Object::create();
```
```suggestion
    Ref newObject = JSON::Object::create();
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -270,18 +270,21 @@ static String processFileAndExtractZipArchive(const String& path)
         recordError(createError(Error::InvalidManifest));
         return false;
     }
+    localization->localizedJSONforJSON(manifestJSON->asObject());
```
```suggestion
    }

    localization->localizedJSONforJSON(manifestJSON->asObject());
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -644,11 +643,11 @@ static WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
         return resourceContents;
 
-    auto* localization = this->localization();
+    auto localization = this->localization();
```
```suggestion
    RefPtr localization = this->localization();
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPILocalizationCocoa.mm:
```diff
@@ -54,7 +54,9 @@
     // Documentation: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/i18n/getMessage
 
-    _WKWebExtensionLocalization *localization = extensionContext().localization();
+    RefPtr<WebExtensionLocalization> localization = extensionContext().localization();
```
```suggestion
    RefPtr localization = extensionContext().localization();
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPILocalizationCocoa.mm:
```diff
@@ -65,7 +67,11 @@
-    return [localization localizedStringForKey:messageName withPlaceholders:substitutionsArray];
+    Vector<String> substitutionsVector;
+    for (id str in substitutionsArray)
+        substitutionsVector.append(str);
```
```suggestion
    for (id string in substitutionsArray)
        substitutionsVector.append(string);
```

xeenon commented on Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionContextProxyCocoa.mm:
```diff
@@ -166,9 +166,17 @@
+
+    if (auto value = JSON::Value::parseJSON(String::fromUTF8(json->span()))) {
+        if (auto obj = value->asObject())
+            return WebExtensionLocalization::create(obj, baseURL.host().toString());
```
```suggestion
    if (RefPtr value = JSON::Value::parseJSON(String::fromUTF8(json->span()))) {
        if (RefPtr object = value->asObject())
            return WebExtensionLocalization::create(object, baseURL.host().toString());
```

xeenon reviewed (changes requested):
Looking good! Just some concern about the RTL/LTR part.

Happy Thanksgiving!

xeenon commented on Source/WebCore/platform/text/PlatformLocale.h:
```diff
@@ -59,6 +59,17 @@ class Locale {
+    enum class WritingDirection: int {
+        WritingDirectionDefault = 0,
+        WritingDirectionLeftToRight = 1,
+        WritingDirectionRightToLeft = 2
```
```suggestion
    enum class WritingDirection: uint8_t {
        Default,
        LeftToRight,
        RightToLeft
```

xeenon commented on Source/WebCore/platform/text/cocoa/LocaleCocoa.mm:
```diff
@@ -145,6 +149,18 @@
+    case NSWritingDirectionRightToLeft:
+        return WritingDirection::WritingDirectionRightToLeft;
+    default:
+        return WritingDirection::WritingDirectionDefault;
```
```suggestion
        return WritingDirection::LeftToRight;
    case NSWritingDirectionRightToLeft:
        return WritingDirection::RightToLeft;
    default:
        return WritingDirection::Default;
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp:
```diff
@@ -0,0 +1,349 @@
+    m_localeString = withBestLocale;
+    m_uniqueIdentifier = uniqueIdentifier;
+
+    RefPtr<JSON::Object> localizationJSON = predefinedMessages(defaultLocalization ?: regionalLocalization ?: languageLocalization);
```
```suggestion
    RefPtr localizationJSON = predefinedMessages();
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp:
```diff
@@ -0,0 +1,349 @@
+    Ref predefinedMessages = JSON::Object::create();
+
+    auto createMessageKey = [](String messageValue) {
+        Ref<JSON::Object> object = JSON::Object::create();
```
```suggestion
        Ref object = JSON::Object::create();
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp:
```diff
@@ -0,0 +1,349 @@
+        return object;
+    };
+
+    if (localizedJSON && m_locale && !m_localeString.isEmpty()) {
```
```suggestion
    if (m_locale && !m_localeString.isEmpty()) {
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp:
```diff
@@ -0,0 +1,349 @@
+    return json ? json->asObject() : nullptr;
+}
+
+Ref<JSON::Object> WebExtensionLocalization::predefinedMessages(RefPtr<JSON::Object> localizedJSON)
```
Param not needed anymore.
```suggestion
Ref<JSON::Object> WebExtensionLocalization::predefinedMessages()
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp:
```diff
@@ -0,0 +1,349 @@
+    m_uniqueIdentifier = uniqueIdentifier;
+
+    RefPtr<JSON::Object> localizationJSON = predefinedMessages(defaultLocalization ?: regionalLocalization ?: languageLocalization);
+    localizationJSON = JSONWithLowercaseKeys(localizationJSON);
```
This is not needed since it is only predefined messages for far, and those are already lowercase.
```suggestion
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionUtilities.cpp:
```diff
@@ -72,6 +72,32 @@ double largestDisplayScale()
     return largestDisplayScale;
 }
 
+RefPtr<JSON::Object> JSONWithLowercaseKeys(RefPtr<JSON::Object> json)
```
```suggestion
RefPtr<JSON::Object> jsonWithLowercaseKeys(RefPtr<JSON::Object> json)
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionUtilities.cpp:
```diff
@@ -72,6 +72,32 @@ double largestDisplayScale()
+        return jsonA ?: jsonB;
+
+    RefPtr mergedObject = jsonA.copyRef();
+    for (String key : jsonB->keys()) {
```
```suggestion
    for (auto& key : jsonB->keys()) {
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionUtilities.cpp:
```diff
@@ -72,6 +72,32 @@ double largestDisplayScale()
+        return json;
+
+    Ref newObject = JSON::Object::create();
+    for (String key : json->keys())
```
```suggestion
    for (auto& key : json->keys())
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPILocalizationCocoa.mm:
```diff
@@ -65,7 +67,11 @@
-    return [localization localizedStringForKey:messageName withPlaceholders:substitutionsArray];
+    Vector<String> substitutionsVector;
+    for (id string in substitutionsArray)
+        substitutionsVector.append(string);
```
Need `#import <wtf/cocoa/VectorCocoa.h>` but this will help:
```suggestion
    auto substitutionsVector = makeVector<String>(substitutionsArray);
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp:
```diff
@@ -0,0 +1,349 @@
+
+void WebExtensionLocalization::loadRegionalLocalization(RefPtr<JSON::Object> regionalLocalization, RefPtr<JSON::Object> languageLocalization, RefPtr<JSON::Object> defaultLocalization, const String& withBestLocale, const String& uniqueIdentifier)
+{
+    m_locale = WebCore::Locale::create(AtomString { withBestLocale });
```
```suggestion
void WebExtensionLocalization::loadRegionalLocalization(RefPtr<JSON::Object> regionalLocalization, RefPtr<JSON::Object> languageLocalization, RefPtr<JSON::Object> defaultLocalization, const String& bestLocale, const String& uniqueIdentifier)
{
    m_locale = WebCore::Locale::create(AtomString { bestLocale });
```

ItsJamie9494 committed:
Port WKWebExtensionLocalization to C++
https://webkit.org/b/283791

Reviewed by Timothy Hatcher.

Create a new WebExtensionLocalization shared class and replace usage of WKWebExtensionLocalization with the new class wherever possible.

* Source/WTF/wtf/JSONValues.h:
* Source/WebKit/Shared/Extensions/WebExtensionContextParameters.h:
* Source/WebKit/Shared/Extensions/WebExtensionContextParameters.serialization.in:
* Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp: Added.
(WebKit::WebExtensionLocalization::WebExtensionLocalization):
(WebKit::WebExtensionLocalization::loadRegionalLocalization):
(WebKit::WebExtensionLocalization::localizedJSONforJSON):
(WebKit::WebExtensionLocalization::localizedStringForKey):
(WebKit::WebExtensionLocalization::localizedArrayForArray):
(WebKit::WebExtensionLocalization::localizedStringForString):
(WebKit::WebExtensionLocalization::localizationJSONForWebExtension):
(WebKit::WebExtensionLocalization::predefinedMessages):
(WebKit::WebExtensionLocalization::stringByReplacingNamedPlaceholdersInString):
(WebKit::WebExtensionLocalization::stringByReplacingPositionalPlaceholdersInString):
* Source/WebKit/Shared/Extensions/WebExtensionLocalization.h: Added.
(WebKit::WebExtensionLocalization::create):
(WebKit::WebExtensionLocalization::uniqueIdentifier):
(WebKit::WebExtensionLocalization::localizationJSON):
(WebKit::WebExtensionLocalization::localizedStringForKey):
(WebKit::WebExtensionLocalization::loadRegionalLocalization):
(WebKit::WebExtensionLocalization::stringByReplacingPositionalPlaceholdersInString):
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.cpp:
(WebKit::JSONWithLowercaseKeys):
(WebKit::mergeJSON):
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.h:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::parseManifest):
(WebKit::WebExtension::serializeLocalization): Deleted.
(WebKit::WebExtension::localization): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::localization):
(WebKit::WebExtensionContext::localizedResourceString):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionURLSchemeHandlerCocoa.mm:
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
(WebKit::WebExtension::serializeLocalization):
(WebKit::WebExtension::localization):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPILocalizationCocoa.mm:
(WebKit::WebExtensionAPILocalization::getMessage):
* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionContextProxyCocoa.mm:
(WebKit::WebExtensionContextProxy::getOrCreate):
(WebKit::WebExtensionContextProxy::parseLocalization):
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:

Canonical link: https://commits.webkit.org/287187@main

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp:
```diff
@@ -0,0 +1,348 @@
+        auto valueType = value->type();
+
+        if (valueType == JSON::Value::Type::String)
+            value->asString() = localizedStringForString(value->asString());
```
@ItsJamie9494 This is incorrect and should be fixed.

Ideally both `localizedJSONforJSON()` and `localizedArrayForArray()` would return a modified copy, and not modify the input.

ItsJamie9494 replied:
to be clear, what's wrong here is modifying the original object instead of storing the localised string in a new object?

ItsJamie9494 replied:
Should be fixed in https://github.com/WebKit/WebKit/pull/37298

xeenon replied:
That way the locale can change and the manifest substitutions could be resolved again in the new language. (We don't do a good job with that currently though.)
